recipes-bsp/acpi-tables/files/edison/acpi-tables-load

```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+# This script copies all ACPI tables into /kernel/firmware/acpi
+# You must have configfs enabled for this to work
+
+    for TABLE in `ls /kernel/firmware/acpi/ 2>/dev/null`; do
+        dest_dir=$(basename $TABLE .aml)
+        mkdir /sys/kernel/config/acpi/table/$dest_dir
+        cat /kernel/firmware/acpi/$TABLE > /sys/kernel/config/acpi/table/$dest_dir/aml
+    done
+
+    gpioset `gpiofind TRI_STATE_ALL`=0
```

Comment:
The problem here that you are setting it **after** loading tables. Moreover, the pin name is not known before loading GPIO expanders. As far as I understand there is no solution for the issue, except exporting the pin and setting it inside *arduino.asl* to "high". The workaround is to create a *gpioexp.asl* without any muxing and do something like

- load *gpioexp*
- set TRI_STATE_ALL to "low"
- load tables
- set TRI_STATE_ALL to "high"

But this is fragile since it relies on current behave of GPIO subsystem (leaving GPIO in the last state, there is a discussion to force them to the state before GPIO request had been done).

Reply:
According to intel docs you need to clear TRI_STATE_ALL, then load tables, then set TRI_STATE_ALL. And as you correctly note TRI_STATE_ALL does not exist before loading the table.

Nevertheless toggling TRI_STATE_ALL is necessary for the GPIO's to work.

I found the order in the script to work. So load tables, clear and set TRI_STATE_ALL. There might be glitches on the GPIO's, I didn't check with the scope. But I think they aren't glitch free on startup anyway.

Reply:
Yes, the problem is that: if TRI_STATE_ALL is not high already, the loading tables would get non-working drivers (because ->probe() functions fail). I saw this many times.  So above flow I mention wouldn't work either. There is no solution at all, because when tables are loaded via ACPI there is no way to defer driver ->probe().

Reply:
OK, the order of scanning is defined somehow (but I'm not sure if it's specified, of course it is for parent-child, but the question about siblings), that's why I put *arduino,asli* as a first part in the resulting _*.asl_. The only way to solve this as I can see is to assert TRI_STATE_ALL in U-Boot and then it wouldn't be different how tables are come into place. The drawback is that hardware will have all possible states on the pins.

Reply:
More I think about it, I come to the following work around:
- specify two identical ASL code with MUXing only (no actual devices described), except one has TRI_STATE_ALL deasserted and the other one asserted.
- specify ASL with the same MUXing, but w/o hogging TRI_STATE_ALL
- do the following: 
 1. Load MUX with TRI_STATE_ALL=0
 2. Load MUX with TRI_STATE_ALL=1
 3. Load normal tables w/o TRI_STATE_ALL hogged.

This would guarantee no glitches on the lines AFAICS.

Reply:
> Yes, the problem is that: if TRI_STATE_ALL is not high already, the loading tables would get non-working drivers (because ->probe() functions fail). I saw this many times. So above flow I mention wouldn't work either. There is no solution at all, because when tables are loaded via ACPI there is no way to defer driver ->probe(). And there is a HW design issue (who the heck invented TRI_STATE_GPIO to be on the same board as all devices we have to gate?!).
> 
> In your case it's "working" because you are actually didn't try complex configurations like Adafruit TFT display.

I thought TRI_STATE_ALL only controls the level translators by setting INH on U5, U7 .. U15 and pulling V_SHIELD_SW low U4, U6 .. U14 and U18 .. U33. But I never really understood how that affects functioning of the muxes. And problems with the muxes after reboot, I haven't seen for a long time, but maybe because I press the RESET longer then before.

Is there no way to define a method in asl that toggles TRI_STATE_ALL on loading the table? Loading/unloading a table sounds like a lot of action. Also, should the correct order not be:
1, 3, 2 (TRI_STATE_ALL goes high after tables loaded)?

Reply:
It affects devices to be recognized. If we have TRI STATE asserted, it means all devices are virtually disconnected and `->probe()` will fail.

The order I put it sounds correct to me. Becaues only at last stage we load actual devices with MUX tables.

And there is no hooks to do anything on loading or unloading tables. This entire concept (of **un**loading tables) is a hack from ACPI(CA) prospective.

Reply:
1. I still haven't figured acpi out. But I read there are notify methods for hot insertion and hot removal? And _INI method (as they write: load when OSPM loads a table).

2. 'Devices to be recognized' you mean on the I2C-6 or SPI bus? I am beginning to understand that the Edison HW guide requires toggling TRI STATE only to reset the devices on these busses after configuring external muxes messes up the bus state. Or actually they want to de-assert during configuring external muxes and assert after. But they are only using MRAA from user space so no ->probe() fails.

3. Your are suggesting TRI_STATE_ALL remains unchanged during warm boot. I did some testing (click on summary  to see results):

 <details>
  <summary>Test results assert/de-assert TRI_STATE_ALL</summary>
*** Boot without tables loaded

```
edison login: root
Last login: Sun Mar 24 19:50:39 UTC 2019 on ttyS2
root@edison:~# modprobe -i i2c-dev
root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
20: 20 21 22 23 -- -- -- -- -- -- -- -- -- -- -- -- 
```

-> not probed

*** Load table

```
root@edison:~# mkdir /sys/kernel/config/acpi/table/arduino
root@edison:~# cat /kernel/firmware/acpi/arduino.aml > /sys/kernel/config/acpi/table/arduino/aml
root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
20: UU UU UU UU -- -- -- -- -- -- -- -- -- -- -- -- 

root@edison:~# i2cdetect -r 6
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- 39 -- -- -- -- -- -- 
```

-> muxes probed, device 39 on I2C-6 detected

*** Deassert TRI_STATE_ALL

```
root@edison:~# gpioset `gpiofind TRI_STATE_ALL`=0
root@edison:~# i2cdetect -r 6
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 

root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: UU UU UU UU -- -- -- -- -- -- -- -- -- -- -- -- 
```

-> mux still probed, device 39 on I2C-6 disappeared (not powered)

*** Reboot with de-asserted TRI_STATE_ALL

```
root@edison:~# reboot usb
edison login: root
root@edison:~# modprobe -i i2c-dev
root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: 20 21 22 23 -- -- -- -- -- -- -- -- -- -- -- -- 

root@edison:~# i2cdetect -r 6

     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
```

-> device 39 on I2C-6 disappeared (not powered), TRI_STATE_ALL remains de-asserted across reboot

*** Load table

```
root@edison:~# mkdir /sys/kernel/config/acpi/table/arduino
root@edison:~# cat /kernel/firmware/acpi/arduino.aml > /sys/kernel/config/acpi/table/arduino/aml
root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
20: UU UU UU UU -- -- -- -- -- -- -- -- -- -- -- -- 

root@edison:~# i2cdetect -r 6
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
```

*** Assert TRI_STATE_AL

```
root@edison:~# gpioset `gpiofind TRI_STATE_ALL`=1
root@edison:~# i2cdetect -r 6
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- 39 -- -- -- -- -- -- 
```

</details> 

These tests show you are right, TRI_STATE_ALL remains de-asserted across reboot.

So, what we should be doing is (note question mark at the end):
- load arduino.aml
- toggle TRI_STATE_ALL
- load tables for i2c-6/spi devices

or use _INI method in i2c-6/spi devices tables to assert TRI_STATE_ALL to power the devices prior to probe()?

Reply:
If `_INI()` works it would be considered as a workaround, indeed. Let me try at some point with my devices. It would need to add an Operation region I suppose to the NIO4, where this pin is stated.

Reply:
Test above was with I2C sensor on i2c-6 bus. Except no kernel driver.

Given we drop the patch for now I guess I can mark this as resolved.